core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -197,6 +198,10 @@ public IndexShard getShard(int shardId) {
         return shards.keySet();
     }
 
+    public Set<ShardRouting> shardRoutings() {
+        return shards.values().stream().map(IndexShard::routingEntry).collect(Collectors.toSet());
```

Comment:
This method is only used in applyDeletedShards to iterate over the shards... That method is called every cluster state update for each IndexService, which might be wasteful in terms of resources. We can simply use iterator instead.